### ISSUE TITLE
Feedback after reviewing KVision

### ISSUE THREAD

**Issue by felixblaschke** [rjaros/kvision#144]
Hi! I reviewed KVision for my IT-company and evaluated a possible further usage. I wanted to share my feedback with you.

First off I am very impressed with this project since it's a one-men-show. There is a well written introduction and ready-to-use templates to start with. (In the beginning I had problems with the live reload cause I started the "run task" gradle build from IntellIJ IDEA. I swapped to terminal command and it worked. Maybe you might want to add this to your documentation)

I looked into your core architecture. It's well designed and structured. Good job.

While structuring and styling my website, KVision felt a bit slow (about 3 seconds per dev-cycle).
Other technologies (Angular, Stencil) are faster here.

I liked the (Flutter-like) approach: UI is code. It's very efficient and clean to have everything in one place. At this point I'd like KVision to be more clear. It covers a large set of basic CSS properties (margin, padding, color, border, etc.) but it has no interface to propper integrate unknown CSS properties. For example I missed out "box-shadow". There was no practical way to add it to an virtual node. I saw that your StyledComponent class support arbitary styles but has no interface to add them. I needed to fallback to write a CSS files, attach a class to my component (while having no CSS-scoping support). This isn't the way of KVision's vision: _"KVision allows you to build modern web applications with the Kotlin language, **without** any use of HTML, CSS or JavaScript."_

In my opinion I would totally focus on **UI is code** and further extend the (code-based) Style support.

Also you might add aliases to h1(), p() or div() since that is HTML terminology.

Another thing I stumbled was the usage of the routing. Your page router correctly triggers the callback, but I needed to manually place, clean and add the page-content. Like this:

```kotlin
val content = div(classes = setOf("content"))
routing
        .on("/", { _ ->
            content.removeAll()
            content.add(mainPage())
        })
        .on("/zwei", { _ ->
            content.removeAll()
            content.add(zweiPage())
        })
        .resolve()
```

A more pratical example in your guide would help here.

In summary I am impressed of your achievement. Keep it up. Just strengthen your "no html/css/js" appraoch even further! ;-)

**Comment by rjaros** [comment]
Thank you for your feedback. I'm really glad you like KVision :-)

I've added recommendation about running build tasks from the command-line to the documentation.

I'll definitely think about supporting arbitrary CSS styles. 

I don't really understand what you mean about aliases, but h1(), p() or div() are already existing names for DSL builder functions.

As for the routing, it's already integrated with `StackPanel` component, so your case could be easily handled using this component. However, in a large, real-life application routing is usually much more complicated (e.g. supports path and url parameters), and direct access to the router's callbacks is very convenient - imho of course :-)

**Comment by nickhristov** [comment]
To add to this, I think its a well designed framework. I like the front-side DSL.

But there is one weak point that the framework does not allow for true extensibility, and some critical things are missing: for example, I cannot create an anchor element, because there is no such Tag in the Tag enum. 

Using an enum here is an extensibility weakness.

**Comment by rjaros** [comment]
I'll think about this as well.

**Comment by rjaros** [comment]
I've added support for both custom HTML elements and arbitrary CSS styles. Since 3.6.0 you can do:
```kotlin
customTag("custom-element") {
    setStyle("box-shadow", "5px 10px")
    p("lorem ipsum")
}
```
